Framework/DataHandling/src/Mantid3MFFileIO.cpp

```diff
@@ -0,0 +1,511 @@
+
+  for (auto i : triangles) {
+    m_triangle.push_back(i.m_Indices[0]);
+    m_triangle.push_back(i.m_Indices[1]);
+    m_triangle.push_back(i.m_Indices[2]);
+  }
+  std::vector<Lib3MF::sPosition> vertices;
+  meshObject->GetVertices(vertices);
+  for (auto i : vertices) {
+    Mantid::Kernel::V3D vertex = createScaledV3D(
+        i.m_Coordinates[0], i.m_Coordinates[1], i.m_Coordinates[2]);
+    m_vertices.push_back(vertex);
+  }
+
+  Mantid::Kernel::Material material;
+
+  Lib3MF_uint32 nResourceID;
+  Lib3MF_uint32 nPropertyID;
+
+  // load the material from the pid\pindex attributes on the mesh object
+  if (meshObject->GetObjectLevelProperty(nResourceID, nPropertyID)) {
+
+    Lib3MF::ePropertyType propType = model->GetPropertyTypeByID(nResourceID);
+
+    if (propType == Lib3MF::ePropertyType::BaseMaterial) {
+
+      Lib3MF::PBaseMaterialGroup baseMaterialGroup =
+          model->GetBaseMaterialGroupByID(nResourceID);
+      ReadMaterial::MaterialParameters params;
+      std::string fullMaterialName = baseMaterialGroup->GetName(nPropertyID);
+      std::string materialName;
+      size_t openBracket = fullMaterialName.find("(");
+      size_t closeBracket = fullMaterialName.find(")");
+      if ((openBracket != std::string::npos) &&
+          (closeBracket != std::string::npos)) {
+        materialName = fullMaterialName.substr(0, openBracket);
+        materialName.erase(
+            std::remove_if(materialName.begin(), materialName.end(), ::isspace),
+            materialName.end());
+        std::string materialSpec = fullMaterialName.substr(
+            openBracket + 1, closeBracket - openBracket - 1);
+        Poco::XML::DOMParser parser;
+        Poco::XML::AutoPtr<Poco::XML::Document> doc;
+        try {
+          doc = parser.parseString("<material id=\"" + materialName +
+                                   "\" formula=\"" + materialName + "\" " +
+                                   materialSpec + "></material>");
+          Poco::XML::AutoPtr<Poco::XML::NodeList> materialElements =
+              doc->getElementsByTagName("material");
+          Kernel::MaterialXMLParser materialParser;
+          material = materialParser.parse(
+              static_cast<Poco::XML::Element *>(materialElements->item(0)));
+        } catch (std::exception) {
+          g_log.warning("Unable to parse material properties for " +
+                        fullMaterialName + " so material will be ignored");
+        }
+      } else {
+        g_log.warning(
+            "Material name " + fullMaterialName +
+            " found without any properties so material will be ignored");
+      }
+    }
+  };
+
+  auto mesh = std::make_shared<Geometry::MeshObject>(
+      std::move(m_triangle), std::move(m_vertices), material);
+  mesh->setID(meshObject->GetName());
+
+  // 3MF stores transformation matrix as row major so transpose
+  // to store in Mantid Matrix class
+
+  // scale the translation entries on 3rd row into metres
```

Comment:
Could you check that this is the simpleist way to transpose a matrix? it is quite bulky, then you could sale the required values seperatly.

Reply:
Good point. I've just pushed a commit that uses some loops to do this rather than explicitly setting up each entry in the matrix